jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,659 @@
+         * {@link Handler.Wrapper Wrapper} class optional.
+         */
+        SKIP_NEXT,
+    }
+
+    private final NotHandled _notHandled;
+    private final IncludeExclude<String> _methods = new IncludeExclude<>();
+    private final IncludeExclude<String> _paths = new IncludeExclude<>(PathSpecSet.class);
+    private final IncludeExcludeSet<Predicate<Request>, Request> _predicates = new IncludeExcludeSet<>(PredicateSet.class);
+    private MethodHandle _doHandle;
+
+    public ConditionalHandler()
+    {
+        this(NotHandled.DO_NOT_HANDLE);
+    }
+
+    public ConditionalHandler(Handler nextHandler)
+    {
+        this(NotHandled.DO_NOT_HANDLE);
+        setHandler(nextHandler);
+    }
+
+    public ConditionalHandler(NotHandled notHandled)
+    {
+        _notHandled = notHandled;
+    }
+
+    /**
+     * Clear all inclusions and exclusions.
+     */
+    public void clear()
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        _methods.clear();
+        _paths.clear();
+        _predicates.clear();
+    }
+
+    /**
+     * Include {@link Request#getMethod() method}s in the conditions to be met
+     * @param methods The exact case-sensitive method name
+     */
+    public void includeMethod(String... methods)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        _methods.include(methods);
+    }
+
+    /**
+     * Exclude {@link Request#getMethod() method}s in the conditions to be met
+     * @param methods The exact case-sensitive method name
+     */
+    public void excludeMethod(String... methods)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        _methods.exclude(methods);
+    }
+
+    /**
+     * Include {@link PathSpec}s in the conditions to be met
+     * @param paths The {@link PathSpec}s that are tested against the {@link Request#getPathInContext(Request) pathInContext}.
+     */
+    public void include(PathSpec... paths)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+
+        for (PathSpec p : paths)
+            ((PathSpecSet)_paths.getIncluded()).add(p);
+    }
+
+    /**
+     * Exclude {@link PathSpec}s in the conditions to be met
+     * @param paths The {@link PathSpec}s that are tested against the {@link Request#getPathInContext(Request) pathInContext}.
+     */
+    public void exclude(PathSpec... paths)
+    {
+        if (isStarted())
+            throw new IllegalStateException(getState());
+        for (PathSpec p : paths)
+            ((PathSpecSet)_paths.getExcluded()).add(p);
+    }
+
+    /**
+     * Include {@link PathSpec}s in the conditions to be met
+     * @param paths String representations of {@link PathSpec}s that are
+     *              tested against the {@link Request#getPathInContext(Request) pathInContext}.
+     */
+    public void includePath(String... paths)
```

Comment:
This should just be an inner class PathPredicate that is (eventually) composited together with the other Predicates for this `ConditionalHandler`

Reply:
No. needed for efficiency.